File: go/pe008.go
// http://projecteuler.net/problem=8
package main

import (
	"fmt"
	"io/ioutil"
	"os"
)

func main() {
	bytes, err := ioutil.ReadFile(os.Args[1])
	if err != nil {
		fmt.Println(err)
		return
	}

	digits := make([]int, 0)
	for _, byte := range bytes {
		byte -= 48
		if byte >= 0 && byte <= 9 {
			digits = append(digits, int(byte))
		}
	}

	max := 0
	for i := 0; i < len(digits)-5; i++ {
		product := 1
		for j := 0; j < 5; j++ {
			product *= digits[i+j]
		}

		if product > max {
			max = product
		}
	}

	fmt.Println(max)
}

File: haskell/pe008.hs
-- | http://projecteuler.net/problem=8
module Euler8 where

import Data.Char (digitToInt, isDigit)
-- import Data.List (tails)

main :: IO ()
main = do
    input <- getContents
    let output = euler8 input
    print output
-- main = getContents >>= print . euler8

euler8 :: String -> Int
euler8 x
    = maximum
    . map product
    . takes 13
    . map digitToInt
    $ filter isDigit x
-- euler8 = maximum . map product . takes 13 . map digitToInt . filter isDigit

takes :: Int -> [a] -> [[a]]
takes _ [] = []
takes n xs@(_ : ys) = take n xs : takes n ys
-- takes n = map (take n) . tails
